worker/src/sidechain_storage/storage.rs

```diff
@@ -0,0 +1,887 @@
+				.map_err(Error::OperationalError)?;
+			self.last_blocks.remove(&shard); // delete from local memory
+
+			// Remove all blocks from db
+			while let Some(previous_block) = self.get_previous_block(shard, last_block.number) {
+				last_block = previous_block;
+				// Blockhash -> Signed Block
+				self.db.delete(last_block.hash.encode()).map_err(Error::OperationalError)?;
+				// (Shard, Block number) -> Blockhash
+				self.db
+					.delete((shard, last_block.number).encode())
+					.map_err(Error::OperationalError)?;
+			}
+			// remove shard from list
+			// STORED_SHARDS_KEY -> Vec<(Shard)>
+			self.shards.retain(|&x| x != *shard);
+			self.db
+				.put(STORED_SHARDS_KEY.encode(), self.shards.encode())
+				.map_err(Error::OperationalError)?
+		}
+		Ok(())
+	}
+
+	/// purges a shard and its block from the db storage
+	pub fn prune_shard_from_block_number(
+		&mut self,
+		shard: &ShardIdentifier,
+		block_number: BlockNumber,
+	) -> Result<()> {
+		if self.shards.contains(&shard) {
+			// get last block of shard
+			let last_block = match self.last_blocks.get(shard) {
+				Some(last_block) => last_block.clone(),
+				None => return Err(Error::LastBlockNotFound(*shard)),
+			};
+			if last_block.number == block_number {
+				// given block number is last block of chain - purge whole shard
+				self.purge_shard(shard)?;
+			} else {
+				// remove given block from db storage
+				let mut current_block_number = block_number;
+
+				// Remove blocks from db until no block anymore
+				while let Some(block_hash) = self.get_block_hash(&shard, current_block_number)? {
+					// Blockhash -> Signed Block
+					self.db.delete(block_hash.encode()).map_err(Error::OperationalError)?;
+					// (Shard, Block number) -> Blockhash
+					self.db
+						.delete((shard, current_block_number).encode())
+						.map_err(Error::OperationalError)?;
+					current_block_number -= 1;
+				}
+			}
+		}
+		Ok(())
+	}
+
+	/// prunes all shards except for the newest blocks (according to blocknumber)
+	pub fn prune_shards(&mut self, blocks_to_keep: BlockNumber) {
+		for shard in self.shards().clone() {
+			// get last block:
+			if let Some(last_block) = self.last_block_of_shard(&shard) {
+				let threshold_block = last_block.number - blocks_to_keep;
+				if let Err(e) = self.prune_shard_from_block_number(&shard, threshold_block) {
+					error!("Could not purge shard {:?} due to {:?}", shard, e);
+				}
+			} else {
+				error!("Last block not found in shard {:?}", shard);
+			}
+		}
+	}
+}
+
+/// implementations of helper functions, not meant for pub use
+impl SidechainStorage {
+	/// gets the previous block of given shard and block number, if there is one
+	fn get_previous_block(
+		&self,
+		shard: &ShardIdentifier,
+		current_block_number: BlockNumber,
+	) -> Option<LastSidechainBlock> {
+		let prev_block_number = current_block_number - 1;
+		if let Some(block_hash_encoded) = self.db.get((shard, prev_block_number).encode()).unwrap()
```

Comment:
Unwraps here intentional?

Reply:
No, it was not, thanks for pointing it out. Fixed :)